### DISCUSSION TITLE
which version of curl supports auth-int for PUT or POST

### DISCUSSION THREAD

**Discussion by yangziok** [curl/curl#7556] [Q&A] [answered]
### I did this
 I find that the library does not support "qop=auth-int" for PUT request,
when I try to send this kind of PUT request, it will fail,but I can send GET request with "qop=auth-int" successfully.
so I want to know, is there any possible to support "qop=auth-int" for PUT request, and whether the latest version supports it.

### I expected the following
 send this kind of PUT request successfully

### curl/libcurl version
curl version 7.70.0

[curl -V output]

### operating system
Linux

**Comment by jay** [comment]
What is the actual `curl -V` version? Have you used [verbose](https://curl.se/docs/manpage.html#-v) mode to check the headers sent by the server? curl prefers auth over auth-int if the server says either is ok. I don't know the reasons for that. If the server says only use auth-int then curl will use it. I'm able to send auth-int PUT requests when I use curl with OpenSSL, but not schannel (Windows SSL). The latter does not support it (SEC_E_QOP_NOT_SUPPORTED).

~~~
curl 7.79.0-DEV (i386-pc-win32) libcurl/7.79.0-DEV OpenSSL/1.1.1j nghttp2/1.43.0
Release-Date: [unreleased]
Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile NTLM SSL Unicode UnixSockets alt-svc
~~~

**Comment by yangziok** [comment]
Hi ,
curl -V output is:
curl 7.70.0 (arm-oe-linux-gnueabi) libcurl/7.70.0 OpenSSL/1.1.1i zlib/1.2.11
Release-Date: 2018-07-11
Protocols: dict file ftp ftps gopher http https imap imaps pop3 pop3s rtsp smbs smb smtp smtps telnet tftp
Features: AsynchDNS Debug TrackMemory IPv6 Largefile NTLM NTLM_WB SSL libz TLS-SRP UnixSockets HTTPS-proxy
------------------------------------------------------------------------------


server says only auth-int in 401 unauthorized response.
SSL is not needed in my environment, so HTTP packets go without any SSL.


I got the latest version of curl code, auth_create_digest_http_message() shows "hash(hashbuf, (const unsigned char *)"", 0);" which is not matched the description like "If the qop value is auth-int, then A2 is:&nbsp; A2 = Method:digest-uri-value:H(entity-body)".
So I want to know that in the version you use, do you modify code of these lines?
if not modify it, does the server give the right 200 OK response after PUT request is sent out?


 




------------------&nbsp;Original&nbsp;------------------
From: &nbsp;"Jay ***@***.***&gt;;
Send time:&nbsp;Wednesday, Aug 11, 2021 11:43 AM
***@***.***&gt;; 
***@***.***&gt;; ***@***.***&gt;; 
Subject: &nbsp;Re: [curl/curl] which version of curl supports auth-int for PUT or POST (#7553)





 
What is the actual curl -V version? Have you used verbose mode to check the headers sent by the server? curl prefers auth over auth-int if the server says either is ok. I don't know the reasons for that. If the server says only use auth-int then curl will use it. I'm able to send auth-int PUT requests when I use curl with OpenSSL, but not schannel (Windows SSL). The latter does not support it (SEC_E_QOP_NOT_SUPPORTED).
 curl 7.79.0-DEV (i386-pc-win32) libcurl/7.79.0-DEV OpenSSL/1.1.1j nghttp2/1.43.0 Release-Date: [unreleased] Protocols: dict file ftp ftps gopher gophers http https imap imaps ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp Features: AsynchDNS Debug HSTS HTTP2 HTTPS-proxy IPv6 Largefile NTLM SSL Unicode UnixSockets alt-svc  
—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by jay** [comment, marked as answer]
According to a comment in the code auth-int is not supported for PUT or POST. It looks like what it is doing is using an empty hash in all cases for the body, so even though it appears to send auth-int for PUT it is probably not actually valid. It is possible if you hash the actual PUT file that it may work, I don't know.
https://github.com/curl/curl/blob/bfbde883af33397943df68a3ae01847a634d33bf/lib/vauth/digest.c#L766-L777
